File: src/lib/prewarm/preprobe-lock.ts
/**
 * Process-local exclusion for speculative pre-rank/probe passes.
 *
 * TorrentFlow is explicitly a single-process local application, so a keyed
 * in-memory lease is the correct boundary. Foreground playback never waits on
 * this lock; only speculative callers contend, and a second pass skips.
 */
const activeUsers = new Set<string>();

export function tryAcquirePreProbeLease(userId: string): (() => void) | null {
  if (activeUsers.has(userId)) return null;
  activeUsers.add(userId);
  let released = false;
  return () => {
    if (released) return;
    released = true;
    activeUsers.delete(userId);
  };
}

export async function tryRunPreProbePass<T>(
  userId: string,
  task: () => Promise<T>,
): Promise<{ started: false } | { started: true; value: T }> {
  const release = tryAcquirePreProbeLease(userId);
  if (!release) return { started: false };
  try {
    return { started: true, value: await task() };
  } finally {
    release();
  }
}

File: server/TorrentFlow.Media/Features/Prewarm/PreProbeLock.cs
namespace TorrentFlow.Media.Features.Prewarm;

/// <summary>
/// Process-local exclusion for speculative pre-rank/probe passes (src/lib/prewarm/preprobe-lock.ts). Foreground
/// playback never waits on this; only speculative callers contend, and a second pass skips.
/// </summary>
public sealed class PreProbeLock
{
    private readonly HashSet<string> _active = new(StringComparer.Ordinal);
    private readonly object _gate = new();

    /// <summary>A release callback (idempotent), or null when a pass for this user is already running.</summary>
    public Action? TryAcquire(string userId)
    {
        lock (_gate)
        {
            if (!_active.Add(userId)) return null;
        }
        var released = 0;
        return () =>
        {
            if (Interlocked.Exchange(ref released, 1) == 1) return;
            lock (_gate) _active.Remove(userId);
        };
    }

    public sealed record PassResult<T>(bool Started, T? Value);

    public async Task<PassResult<T>> TryRunPassAsync<T>(string userId, Func<Task<T>> task)
    {
        var release = TryAcquire(userId);
        if (release is null) return new PassResult<T>(false, default);
        try
        {
            return new PassResult<T>(true, await task());
        }
        finally
        {
            release();
        }
    }
}
